rslib/src/collection/backup.rs

```diff
@@ -0,0 +1,343 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::{
+    collections::HashMap,
+    ffi::OsStr,
+    fs::{self, read_dir, remove_file, DirEntry, File},
+    io::{Read, Write},
+    path::{Path, PathBuf},
+    thread,
+};
+
+use chrono::prelude::*;
+use itertools::Itertools;
+use log::error;
+use serde_derive::{Deserialize, Serialize};
+use zip::{read::ZipFile, write::FileOptions, CompressionMethod, ZipArchive, ZipWriter};
+use zstd;
+
+use crate::{
+    collection::CollectionBuilder,
+    config::BackupLimits,
+    error::{DbError, DbErrorKind},
+    log,
+    prelude::*,
+    text::normalize_to_nfc,
+};
+
+/// Bump if making changes that break restoring on older releases.
+/// Versions 1 and 2 wrote different archives, so minimum expected value is 3.
+const BACKUP_VERSION: u8 = 3;
+const BACKUP_FORMAT_STRING: &str = "backup-%Y-%m-%d-%H.%M.%S.colpkg";
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+#[serde(default)]
+struct Meta {
+    #[serde(rename = "ver")]
+    version: u8,
+}
+
+pub fn backup<P1, P2>(col_path: P1, backup_folder: P2, limits: BackupLimits) -> Result<()>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path> + Send + 'static,
+{
+    let col_file = File::open(col_path)?;
+    let col_data = zstd::encode_all(col_file, 0)?;
+
+    thread::spawn(move || backup_inner(&col_data, &backup_folder, limits));
+
+    Ok(())
+}
+
+pub fn restore_backup(col_path: &str, backup_path: &str, media_folder: &str) -> Result<()> {
+    let backup_file = File::open(backup_path)?;
+    let mut archive = ZipArchive::new(backup_file)?;
+    let new_col_data = extract_collection_data(&mut archive)?;
+    let old_col_data = fs::read(col_path)?;
+    fs::write(col_path, new_col_data)?;
```

Comment:
This approach has a couple of race conditions:

- if the new data fails the check and Anki or the system crashes, the user may be stuck with the new data. Likewise if a crash happens after it's written but before a check is done.
- fs::write() is not atomic, so if a crash happens before all data is written, you can end up with partially written file.

Reading the old file into memory first also increases our memory usage here.

What I'd recommend doing instead is to use NamedTemporaryFile::new_in() to write a temporary file in the same folder as the existing collection (can't use standard temp location, as cross-drive/cross-partition will break things). We can run the tests on that temporary file, and if anything goes wrong, the named file can juts be dropped. Once we're ready to use it, the original file can be replaced atomically via https://docs.rs/tempfile/latest/tempfile/struct.NamedTempFile.html#method.persist, ensuring it doesn't get half-written.

We can probably get away with reading the entire new collection into memory as we are currently doing with decode_all(), but it's not ideal - at one point it might be nice to tweak the code so the collection is decoded and written into the temporary file in chunks, instead of loading everything into memory.